### ISSUE TITLE
BUG: Memory leak when setting Series value via __setitem__

### ISSUE THREAD

**Issue by jzombi** [pandas-dev/pandas#47172]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
import gc
import os
import psutil

# create example data
items = [f'item_{i}' for i in range(10_000)]
data = {}
for i, col in enumerate(items[:1000]):
    data[col] = [1]*len(items)
df = pd.DataFrame(index=items, data=data)
gc.collect()


process = psutil.Process(os.getpid())
rss = process.memory_info().rss/1024/1024

for item in df.columns[:10]:
    df[item][item] = -10
    gc.collect()
    new_rss = process.memory_info().rss/1024/1024
    print('{:.2f} MiB'.format(new_rss-rss))
```

### Issue Description

In pandas version `1.4.2` the memory usage increases at each iteration of the last for loop, outputs:

<details>

    76.55 MiB
    152.99 MiB
    229.16 MiB
    305.08 MiB
    381.00 MiB
    456.92 MiB
    532.58 MiB
    608.24 MiB
    683.96 MiB
    759.51 MiB

</details>

### Expected Behavior

In pandas version `1.3.5` the memory usage remains constant, output:

<details>

    0.00 MiB
    0.00 MiB
    0.00 MiB
    0.00 MiB
    0.00 MiB
    0.00 MiB
    0.00 MiB
    0.00 MiB
    0.00 MiB
    0.00 MiB

</details>

### Installed Versions

<details>

    INSTALLED VERSIONS
    ------------------
    commit           : 4bfe3d07b4858144c219b9346329027024102ab6
    python           : 3.8.12.final.0
    python-bits      : 64
    OS               : Linux
    OS-release       : 5.10.0-13-amd64
    Version          : #1 SMP Debian 5.10.106-1 (2022-03-17)
    machine          : x86_64
    processor        :
    byteorder        : little
    LC_ALL           : None
    LANG             : en_US.UTF-8
    LOCALE           : en_US.UTF-8

    pandas           : 1.4.2
    numpy            : 1.21.6
    pytz             : 2022.1
    dateutil         : 2.8.2
    pip              : 22.1.1
    setuptools       : 56.0.0
    Cython           : 0.29.30
    pytest           : None
    hypothesis       : None
    sphinx           : 4.5.0
    blosc            : None
    feather          : None
    xlsxwriter       : None
    lxml.etree       : None
    html5lib         : None
    pymysql          : 1.0.2
    psycopg2         : None
    jinja2           : 3.1.2
    IPython          : None
    pandas_datareader: None
    bs4              : None
    bottleneck       : None
    brotli           : None
    fastparquet      : None
    fsspec           : None
    gcsfs            : None
    markupsafe       : 2.1.1
    matplotlib       : 3.5.1
    numba            : 0.55.1
    numexpr          : None
    odfpy            : None
    openpyxl         : None
    pandas_gbq       : None
    pyarrow          : None
    pyreadstat       : None
    pyxlsb           : None
    s3fs             : None
    scipy            : 1.8.0
    snappy           :
    sqlalchemy       : 1.4.36
    tables           : None
    tabulate         : None
    xarray           : None
    xlrd             : None
    xlwt             : None
    zstandard        : None

</details>

**Comment by phofl** [comment]
Before 1.4 setitem probably wrote into the existing array, since 1.4 this always makes a copy. Could you try this without chained indexing?

**Comment by phofl** [comment]
I meant using loc, e.g. ``df.loc[item, item]`` (I think this does what you want)

**Comment by phofl** [comment]
The internal cache is not updated correctly when using chained indexing (since 1.4). I think we already have similar issues about that, could you check?

**Comment by jzombi** [comment]
Yes, both:

    df.loc[item, item] = -10

and

    df[item].loc[item] = -10

works fine. Actually I've used the solution you suggested to solve my issue before opening this bugreport. I've reported this as a bug only because I think that changing the values of the dataframe this way, even if it makes a copy, should not increase the memory usage further and further (seems like references are stuck somewhere).

**Comment by jzombi** [comment]
> The internal cache is not updated correctly when using chained indexing (since 1.4). I think we already have similar issues about that, could you check?

I'm sorry, I'm afraid I don't know enough about pandas to recognize which issues are really related.

Another strange thing, if `df.memory_usage()` is called before the for loop, the memory usage does not increase at each iteration step.

    df.memory_usage()
    for item in df.columns[:10]:
        df[item][item] = -10

        gc.collect()
        new_rss = process.memory_info().rss/1024/1024
        print('{:.2f} MiB'.format(new_rss-rss))

**Comment by simonjayhawkins** [comment]
> > Before 1.4 setitem probably wrote into the existing array, since 1.4 this always makes a copy.
> 
> I think so too (maybe related: #43406?)

can confirm.

first bad commit: [03dd698bc1e84c35aba8b51bdd45c472860b9ec3] BUG: DataFrame.__setitem__ sometimes operating inplace (#43406)

cc @jbrockmendel

**Comment by simonjayhawkins** [comment]
> > The internal cache is not updated correctly when using chained indexing (since 1.4). I think we already have similar issues about that, could you check?
> 
> I'm sorry, I'm afraid I don't know enough about pandas to recognize which issues are really related.

several issue point back to #43406. The cache issue may have been referring to is #45684 which should be fixed.

**Comment by simonjayhawkins** [comment]
moving to 1.4.4

**Comment by simonjayhawkins** [comment]
closed in #48215
